storage/local/frankenstein.go

```diff
@@ -98,21 +103,43 @@ func NewIngestor(chunkStore ChunkStore) (*Ingestor, error) {
 			Help:      "The total number of samples returned from queries.",
 		}),
 	}
-	var err error
-	i.mapper, err = newFPMapper(i.fpToSeries, noopPersistence{})
-	if err != nil {
-		return nil, err
-	}
 
 	go i.loop()
 	return i, nil
 }
 
-func (*Ingestor) NeedsThrottling() bool {
+func (i *Ingestor) getStateFor(ctx context.Context) (*userState, error) {
+	userID, ok := ctx.Value(UserIDContextKey).(string)
+	if !ok {
+		return nil, fmt.Errorf("No user id")
+	}
+
+	i.userStateLock.Lock()
+	state, ok := i.userState[userID]
+	if !ok {
+		state = &userState{
+			userID:     userID,
+			fpToSeries: newSeriesMap(),
+			fpLocker:   newFingerprintLocker(16),
+			index:      newInvertedIndex(),
+		}
+		var err error
+		state.mapper, err = newFPMapper(state.fpToSeries, noopPersistence{})
+		if err != nil {
+			return nil, err
+		}
+		i.userState[userID] = state
+	}
+	i.userStateLock.Unlock()
```

Comment:
Since it's happening at the end of the function, better to unlock it via `defer` right after taking the lock.